File: Simulation and Modelling/lab1.py
import numpy as np

# Step 1: Generate 100 Random Numbers using Arithmetic Congruential Method
X0, X1 = 7, 11  # Seed values
m = 31  # Modulus value
N = 100  # Number of random numbers

# Initialize the sequence
X = [X0, X1]

# Generate 98 more random integers
for i in range(2, N):
    Xi = (X[i-1] + X[i-2]) % m
    X.append(Xi)

# Compute the random numbers in the range [0,1]
R = [xi / m for xi in X]

# Step 2: Chi-Square Test for Uniformity
n_intervals = 10  # Number of class intervals
intervals = np.linspace(0, 1, n_intervals + 1)

# Count observed frequencies
observed_freq, _ = np.histogram(R, bins=intervals)

# Expected frequency assuming uniform distribution
expected_freq = [N / n_intervals] * n_intervals

# Compute Chi-Square statistic
chi_square_stat = sum(((observed_freq[i] - expected_freq[i]) ** 2) / expected_freq[i] for i in range(n_intervals))

# Critical value from lab sheet
chi_square_critical = 16.9  

# Step 3: Kolmogorov-Smirnov Test
# Sort the random numbers
R_sorted = sorted(R)

# Compute D+ and D-
D_plus = [(i + 1) / N - R_sorted[i] for i in range(N)]
D_minus = [R_sorted[i] - i / N for i in range(N)]

# Determine D_max
D_max = max(max(D_plus), max(D_minus))

# Critical value for KS test
D_critical = 1.36 / np.sqrt(N)

# Display Results
print("Generated Random Numbers:", R[:10], "...")  # Display first 10 as a preview
print("\nChi-Square Test Results:")
print("Observed Frequencies:", observed_freq)
print("Expected Frequency per interval:", expected_freq[0])
print("Chi-Square Statistic:", chi_square_stat)
print("Critical Value:", chi_square_critical)
print("Conclusion:", "Reject H0 (Not Uniform)" if chi_square_stat > chi_square_critical else "Accept H0 (Uniform)")

print("\nKolmogorov-Smirnov Test Results:")
print("D_max:", D_max)
print("Critical Value:", D_critical)
print("Conclusion:", "Reject H0 (Not Uniform)" if D_max > D_critical else "Accept H0 (Uniform)")

File: Simulation and Modelling/lab1.cpp
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

#define N 100  // Number of random numbers
#define INTERVALS 10  // Number of class intervals (for Chi-Square test)

// Function to generate 100 random numbers in range [0,1]
void generate_random_numbers(double numbers[]) {
    for (int i = 0; i < N; i++) {
        numbers[i] = (double)rand() / RAND_MAX;
    }
}

// Function to sort numbers in ascending order (Bubble Sort for simplicity)
void sort_numbers(double numbers[]) {
    for (int i = 0; i < N - 1; i++) {
        for (int j = i + 1; j < N; j++) {
            if (numbers[i] > numbers[j]) {
                double temp = numbers[i];
                numbers[i] = numbers[j];
                numbers[j] = temp;
            }
        }
    }
}

// Kolmogorov-Smirnov Test
void kolmogorov_smirnov_test(double numbers[]) {
    sort_numbers(numbers);  // Sort the random numbers first

    double d_plus[N], d_minus[N];
    double d_plus_max = 0.0, d_minus_max = 0.0;

    // Compute D+ and D-
    printf("\nKolmogorov-Smirnov Test Table:\n");
    printf("i\tRandom Number\tF_observed (i/N)\tD+\t\tD-\n");

    for (int i = 0; i < N; i++) {
        double i_n = (double)(i + 1) / N;  // i/N
        double i_minus_1_n = (double)i / N;  // (i-1)/N

        d_plus[i] = i_n - numbers[i];   // D+ = i/N - R_i
        d_minus[i] = numbers[i] - i_minus_1_n;  // D- = R_i - (i-1)/N

        if (d_plus[i] > d_plus_max) d_plus_max = d_plus[i];
        if (d_minus[i] > d_minus_max) d_minus_max = d_minus[i];

        printf("%d\t%.4f\t\t%.4f\t\t%.4f\t%.4f\n", i + 1, numbers[i], i_n, d_plus[i], d_minus[i]);
    }

    // Compute D0 = max(D+ max, D- max)
    double d0 = fmax(d_plus_max, d_minus_max);

    // Compute critical value: D_0.05,100 = 1.36 / sqrt(N)
    double critical_value = 1.36 / sqrt(N);

    printf("\nD+ max = %.4f, D- max = %.4f\n", d_plus_max, d_minus_max);
    printf("D0 = max(D+ max, D- max) = %.4f\n", d0);
    printf("Critical Value (D_0.05,100) = %.4f\n", critical_value);

    // Compare D0 with critical value
    if (d0 < critical_value) {
        printf("Result: The random numbers follow a uniform distribution (Fail to Reject H0)\n");
    } else {
        printf("Result: The random numbers do NOT follow a uniform distribution (Reject H0)\n");
    }
}

// Chi-Square Test
void chi_square_test(double numbers[]) {
    int observed_freq[INTERVALS] = {0};
    double expected_freq = (double)N / INTERVALS;  // E[i] = N/n

    // Count occurrences in each interval
    for (int i = 0; i < N; i++) {
        int index = (int)(numbers[i] * INTERVALS);  // Find interval index
        if (index == INTERVALS) index = INTERVALS - 1;  // Handle edge case
        observed_freq[index]++;
    }

    // Compute Chi-Square test statistic
    double chi_square = 0.0;
    printf("\nChi-Square Test Table:\n");
    printf("Interval\tObserved Freq (O_i)\tExpected Freq (E_i)\t(O_i - E_i)^2 / E_i\n");

    for (int i = 0; i < INTERVALS; i++) {
        double diff = observed_freq[i] - expected_freq;
        double term = (diff * diff) / expected_freq;
        chi_square += term;

        printf("[%0.1f, %0.1f]\t%d\t\t%.2f\t\t%.4f\n", 
                (double)i / INTERVALS, (double)(i + 1) / INTERVALS, 
                observed_freq[i], expected_freq, term);
    }

    // Given critical value: X^2_0.05,9 = 16.9
    double chi_critical = 16.9;

    printf("\nChi-Square Test Statistic: X^2_0 = %.4f\n", chi_square);
    printf("Critical Value (X^2_0.05,9) = %.2f\n", chi_critical);

    // Compare test statistic with critical value
    if (chi_square < chi_critical) {
        printf("Result: The random numbers follow a uniform distribution (Fail to Reject H0)\n");
    } else {
        printf("Result: The random numbers do NOT follow a uniform distribution (Reject H0)\n");
    }
}

int main() {
    double numbers[N];

    // Seed the random number generator
    srand(time(NULL));

    // Generate 100 random numbers
    generate_random_numbers(numbers);

    // Print generated numbers
    printf("Generated Random Numbers:\n");
    for (int i = 0; i < N; i++) {
        printf("%.2f ", numbers[i]);
        if ((i + 1) % 10 == 0) printf("\n");
    }

    // Perform Kolmogorov-Smirnov Test
    kolmogorov_smirnov_test(numbers);

    // Perform Chi-Square Test
    chi_square_test(numbers);

    return 0;
}

